### PR TITLE
Add metatables to lua vectors

### PR DESCRIPTION
**Pull request by Desour**
- This PR is basically the same as #10330.
- Goal of the PR: Add metatables to vectors, so you can more easily program with them.
- Included features:
  - All the builtin helpers and also minetest callbacks (like `can_dig`) give vectors with metatables.
  - `:`-syntax: `local a = b:offset(0,1,0)`
  - operators: `local a = -b * 2`
  - indexing with numbers: `local a = v[1]`
  - Many unittests.
- backwards compatibility:
  - There are no weird hacks for keeping backwards compatibility.
  - If you want to use `==`, `:` or the number-indexing on vectors that you've got from other mods, you will probably have to copy them via `vector.new` first.
  - Old code should still work fine, with very few exceptions, eg.: mods that store something in indices `1`..`3` in vectors generated by vector helpers, and mods that depend on vectors generated by vector helpers to have no metatable (something like `assert(getmetatable(v)==nil)`. I highly doubt that such mods are common.
- usecases? Copied from #9612:

old code | new code
----------- | ------
`vector.multiply(vector.add(vector.multiply(a, -1), b), 2)` | `(-a + b) * 2`
`vector.length(a)` | `a:length()`
`local xyz = {"x", "y", "z"} for i = 1, 3 do if v[xyz[i]] == 0 then print("bla") end end` | `for i = 1, 3 do if v[i] == 0 then print("bla") end end`
`if vector.equal(a, b) then` | `if a == b then`
`print(vector.to_string(v))` | `print(v)`

### Performance

TL;DR:
- Imo it's ok. :green_square: 
- At least in luajit if you do stuff in a long loop, the performance of the old operations is not worse.
- The number-indexing is slower.
- In certain situations the new stuff can be faster.

<details>

I've compared master with this PR, using luajit2.1, CPU was clocked at 1600MHz. 

#### code:
```lua

local insec_env = minetest.request_insecure_environment()
--~ local jdump = insec_env.require("jit.dump")
local jdump = insec_env.require("jit.v")

--~ local is_new = true
local is_new = false

local function meassure(name, func, new_stuff_needed, ...)
	if new_stuff_needed and not is_new then
		return
	end
	print(name..":")
	local time_start = minetest.get_us_time()
	--~ jdump.on()
	func(...)
	--~ jdump.off()
	local time_end = minetest.get_us_time()
	print("time taken: " .. ((time_end - time_start) / 1000) .. " ms\n")
end

local is_luajit = rawget(_G, "jit") ~= nil

print("benchmarking...")
print("Lua: " .. (is_luajit and jit.version or "PUC Lua"))
print("")

local n1 = is_luajit and 2000000 or 500000
if false then
	meassure("vec_new_i1", function()
		local v1s = {}
		local v2s = {}
		local v3s = {}
		for i = 1, n1 do
			local v = vector.new(i, i, i)
			--~ local v = vector.from(i, i, i)
			v1s[i] = v
			--~ v2s[i] = vector.copy(v)
			--~ v3s[i] = vector.zero()
			v2s[i] = vector.new(v)
			v3s[i] = vector.new()
		end
		print(#v1s)
		print(#v2s)
		print(#v3s)
	end, true)
end

local n3 = is_luajit and 1000000 or 5000000
local rand_vecs = {}
meassure("create_rand_vecs", function()
	for i = 0, n3 do
		rand_vecs[i] = vector.new(math.random(), math.random(), math.random())
	end
end)

local n2 = is_luajit and 100000000 or 5000000
meassure("vec_add_loop1_.add", function()
	local sum = vector.new()
	for i = 1, n2 do
		sum = vector.add(sum, rand_vecs[i%n3])
	end
	print(sum.x)
end)
meassure("vec_add_loop2_:add", function()
	local sum = vector.new()
	for i = 1, n2 do
		sum = sum:add(rand_vecs[i%n3])
	end
	print(sum.x)
end, true)
meassure("vec_add_loop3_+", function()
	local sum = vector.new()
	for i = 1, n2 do
		sum = sum + rand_vecs[i%n3]
	end
	print(sum.x)
end, true)
meassure("vec_add_loop1_.add", function()
	local sum = vector.new()
	for i = 1, n2 do
		sum = vector.add(sum, rand_vecs[i%n3])
	end
	print(sum.x)
end)
meassure("vec_add_loop1_.add", function()
	local sum = vector.new()
	for i = 1, n2 do
		sum = vector.add(sum, rand_vecs[i%n3])
	end
	print(sum.x)
end)

local n21 = is_luajit and 10000000 or 500000
meassure("vec_add2_loop1_.add", function()
	local sum = vector.new()
	for i = 1, n21 do
		local k = (i+1)%n3
		rand_vecs[k] = vector.add(rand_vecs[k], rand_vecs[i%n3])
	end
	print(sum.x)
end)
meassure("vec_add2_loop2_:add", function()
	local sum = vector.new()
	for i = 1, n21 do
		local k = (i+1)%n3
		rand_vecs[k] = rand_vecs[k]:add(rand_vecs[i%n3])
	end
	print(sum.x)
end, true)
meassure("vec_add2_loop3_+", function()
	local sum = vector.new()
	for i = 1, n21 do
		local k = (i+1)%n3
		rand_vecs[k] = rand_vecs[k] + rand_vecs[i%n3]
	end
	print(sum.x)
end, true)
meassure("vec_add2_loop1_.add", function()
	local sum = vector.new()
	for i = 1, n21 do
		local k = (i+1)%n3
		rand_vecs[k] = vector.add(rand_vecs[k], rand_vecs[i%n3])
	end
	print(sum.x)
end)

local n4 = is_luajit and 100000000 or 5000000
meassure("vec_index_loop1_.x", function()
	--~ jdump.on()
	local sum = 0
	for i = 1, n4 do
		sum = sum + rand_vecs[i%n3].x
		sum = sum + rand_vecs[i%n3].y
		sum = sum + rand_vecs[i%n3].z
	end
	--~ jdump.off()
	print(sum)
end)

meassure("vec_index_loop2_[1]", function()
	local sum = 0
	for i = 1, n4 do
		sum = sum + rand_vecs[i%n3][1]
		sum = sum + rand_vecs[i%n3][2]
		sum = sum + rand_vecs[i%n3][3]
	end
	print(sum)
end, true)

print("benchmarking done.")

```

#### Results:

master:
```
benchmarking...
Lua: LuaJIT 2.1.0-beta3

create_rand_vecs:
time taken: 271.747 ms

vec_add_loop1_.add:
50006693.021157
time taken: 2292.325 ms

vec_add_loop1_.add:
50006693.021157
time taken: 2279.754 ms

vec_add_loop1_.add:
50006693.021157
time taken: 2300.492 ms

vec_add2_loop1_.add:
0
time taken: 2548.892 ms

vec_add2_loop1_.add:
0
time taken: 2057.055 ms

vec_index_loop1_.x:
6.8982152930833e+109
time taken: 1709.975 ms

benchmarking done.
```

this PR:
```
benchmarking...
Lua: LuaJIT 2.1.0-beta3

create_rand_vecs:
time taken: 305.866 ms

vec_add_loop1_.add:
49991247.564712
time taken: 2245.698 ms

vec_add_loop1_.add:
49991247.564712
time taken: 2298.445 ms

vec_add_loop1_.add:
49991247.564712
time taken: 2304.906 ms

vec_add2_loop1_.add:
0
time taken: 2319.513 ms

vec_add2_loop1_.add:
0
time taken: 2274.248 ms

vec_index_loop1_.x:
6.9058452003397e+109
time taken: 1663.239 ms

benchmarking done.
```
also this PR, but all benchmarks enabled:
```
benchmarking...
Lua: LuaJIT 2.1.0-beta3

create_rand_vecs:
time taken: 312.248 ms

vec_add_loop1_.add:
49964549.735267
time taken: 2248.874 ms

vec_add_loop2_:add:
49964549.735267
time taken: 1982.679 ms

vec_add_loop3_+:
49964549.735267
time taken: 1954.441 ms

vec_add_loop1_.add:
49964549.735267
time taken: 2237.673 ms

vec_add_loop1_.add:
49964549.735267
time taken: 2220.716 ms

vec_add2_loop1_.add:
0
time taken: 2404.558 ms

vec_add2_loop2_:add:
0
time taken: 2381.15 ms

vec_add2_loop3_+:
0
time taken: 2024.108 ms

vec_add2_loop1_.add:
0
time taken: 2228.035 ms

vec_index_loop1_.x:
5.2423356774124e+208
time taken: 1663.82 ms

vec_index_loop2_[1]:
5.2423356774124e+208
time taken: 3235.125 ms

benchmarking done.
```

#### evaluation:
- `vec_add_loop1_.add` is slower than the `+` and `:add` versions. I guess this is because we operate on the local sum variable. This was also the reason why I added the `add2_loop`s.
- `vec_add2_loop2_:add` seems to be slower than `vec_add2_loop3_+`. But if I first execute the `+`, then both times are equal `¯\_(ツ)_/¯`. It looks like there's no big performance difference.
- `vec_index_loop1_.x` is much faster than `vec_index_loop2_[1]`.
- Note that I've executed these benchmarks multiple times and the results were always similar. However, one might be able to see some tiny differences by doing the benchmarks much more often and averaging the results. Results might also differ for different lua/-jit versions. I am too lazy to do better benchmarks. If you really need more performance (unlikely), you should consider not using minetest's vector helpers at all

</details>

## To do

This PR is a Ready for Review.

## How to test

- Run the unittests.
- Play with mods that use the `vector.*` helpers, and look for errors.
- Look through the code.
- Do your own performance tests.

### CODE REVIEW

**Comment by Desour** [comment]
It turned out that much of the builtin code constructed vectors using `{x=...}`, which eg. caused the `pos` in `minetest.register_on_dignode`'s callback not to have a metatable. These problems should be fixed now.

**Review by v-rob — COMMENTED** [review]
Nice PR to have; metatable form is much nicer to use.  Just some mostly trivial questions/suggestions.

How do I run the unit tests?  I will look over them and run them after I look at the other vector PRs.

File: doc/lua_api.txt [hunk 1]
```diff
@@ -3190,14 +3215,30 @@ For the following functions `x` can be either a vector or a number:
...
     * Deprecated: If `s` is a vector: Returns the Schur quotient.
 
+Operators can be used if all of the involved vectors have metatables:
+* `v1 == v2`:
+    * Returns `true` iff `v1` and `v2` are identical.
+* `-v`:
+    * Returns the additive inverse of v.
+* `v1 + v2`:
+    * Returns the sum of both vectors.
+    * Note: `+` can not be used together with scalars.
```
[the remark this answers is not in the document]
Comment (Desour) [reply doc/lua_api.txt:3225]: You normally don't add a scalar and a vector in maths, and imo one would expect + to work like the mathematical + on vectors.

File: builtin/common/vector.lua [hunk 2]
```diff
@@ -1,58 +1,84 @@
...
+-- only called when rawget(v, key) returns nil
+function metatable.__newindex(v, key, value)
+	rawset(v, xyz[key] or key, value)
+end
+
+-- constructors
+
+local function fast_new(x, y, z)
+	local v = setmetatable({x = x, y = y, z = z}, metatable)
+	return v
```
[the remark this answers is not in the document]
Comment (Desour) [reply builtin/common/vector.lua:27]: The tail call there was slower once. I currently can't reproduce, though.

File: builtin/common/vector.lua [hunk 3]
```diff
@@ -1,58 +1,84 @@
...
 	       a.z == b.z
 end
+metatable.__eq = vector.equals
+
+-- unary operations
 
 function vector.length(v)
 	return math.hypot(v.x, math.hypot(v.y, v.z))
 end
+-- Note: we can not use __len because it is already used for primitive table length
```
[the remark this answers is not in the document]
Comment (Desour) [reply builtin/common/vector.lua:53]: Are you suggesting to remove that comment?

File: doc/lua_api.txt [hunk 4]
```diff
@@ -3171,7 +3193,10 @@ For the following functions, `v`, `v1`, `v2` are vectors,
 * `vector.cross(v1, v2)`:
     * Returns the cross product of `v1` and `v2`.
 * `vector.offset(v, x, y, z)`:
-    * Returns the sum of the vectors `v` and `{x = x, y = y, z = z}`.
+    * Returns the sum of the vectors `v` and `(x, y, z)`.
+* `vector.is_vector(v)`:
+    * Returns `true` iff `v` is a real vector, eg. created by a `vector.*` function.
```
[the remark this answers is not in the document]
Comment (Desour) [reply doc/lua_api.txt:3198]: `has_metatable` would be too long imo, and doesn't convey what this function is intended to be used for.
But you might be right, a different name might be better.

The "iff"s are on purpose, see https://en.wikipedia.org/wiki/If_and_only_if.

File: doc/lua_api.txt [hunk 5]
```diff
@@ -1482,6 +1482,9 @@ Position/vector
 
     {x=num, y=num, z=num}
 
+    Note: it is highly recommended to construct a vector using the helper function:
+    vector.new(num, num, num)
+
```
[the remark this answers is not in the document]
Comment (Desour) [reply doc/lua_api.txt:1487]: I'd like to keep it:
- You will still see code using the old format.
- All the minetest functions and even the `vector.*` functions accept the raw table.

File: builtin/common/vector.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (v-rob) [reply builtin/common/vector.lua:53]: It doesn't matter whether you remove it or not; I was just commenting that using `#` for vector length goes against the semantics of the operator and isn't vector notation either.

File: doc/lua_api.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (v-rob) [reply doc/lua_api.txt:3225]: Makes sense; I guess it's fine for the function to have more behaviour.  Operators should be obvious to understand without reading the docs.

File: doc/lua_api.txt [hunk 6]
```diff
@@ -3134,25 +3137,44 @@ no particular point.
 
 Internally, it is implemented as a table with the 3 fields
 `x`, `y` and `z`. Example: `{x = 0, y = 1, z = 0}`.
+However, one should *never* create a vector manually as above. The vector helpers
```
[the remark this answers is not in the document]
Comment (Desour) [reply doc/lua_api.txt:3140]: Done.

File: doc/lua_api.txt [hunk 7]
```diff
@@ -3190,14 +3215,30 @@ For the following functions `x` can be either a vector or a number:
     * Returns a scaled vector.
     * Deprecated: If `s` is a vector: Returns the Schur quotient.
 
+Operators can be used if all of the involved vectors have metatables:
+* `v1 == v2`:
+    * Returns `true` iff `v1` and `v2` are identical.
```
[the remark this answers is not in the document]
Comment (Desour) [reply doc/lua_api.txt:3220]: Same as above.

File: src/script/common/c_converter.cpp [hunk 8]
```diff
@@ -51,6 +51,31 @@ if (value < F1000_MIN || value > F1000_MAX) { \
...
 
+/**
+ * A helper which sets (if available) the vector metatable from builtin as metatable
+ * for the table on top of the stack
+ */
+static void set_vector_metatable(lua_State *L)
+{
+	// get vector.metatable
+	lua_pushstring(L, "vector");
+	lua_rawget(L, LUA_GLOBALSINDEX);
```
[the remark this answers is not in the document]
Comment (Desour) [reply src/script/common/c_converter.cpp]: Ok, done.

File: src/script/common/c_converter.cpp [hunk 9]
```diff
@@ -51,6 +51,31 @@ if (value < F1000_MIN || value > F1000_MAX) { \
...
+	if (!lua_istable(L, -1)) {
+		// there is no global vector table
+		lua_pop(L, 1);
+		errorstream << "set_vector_metatable in c_converter.cpp: " <<
+				"missing global vector table" << std::endl;
+		return;
+	}
+	lua_pushstring(L, "metatable");
+	lua_rawget(L, -2);
```
[the remark this answers is not in the document]
Comment (Desour) [reply src/script/common/c_converter.cpp]: `-1`*
Done.

**Comment by Desour** [comment]
Thanks for the review!

>How do I run the unit tests?

- Install busted (eg. with luarocks).
- Run `$ busted builtin` in your minetest dir.

See also: https://github.com/minetest/minetest/pull/9184#issue-351096122

File: doc/lua_api.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (Thomas--S) [reply doc/lua_api.txt:3199]: The word "iff" might indeed be a little bit confusing as it probably seems to be a bit uncommon.

Would it be a viable option to rephrase the sentence in order to avoid this expression?

Suggestion:
```suggestion
    * Returns a boolean value indicating whether `v` is a real vector, eg. created by a `vector.*` function.
```

[the remark this answers is not in the document]
Comment (Desour) [reply doc/lua_api.txt]: Ok, I've applied you suggestion.
It would be simpler if everyone learned to use and read iff.

File: doc/lua_api.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (Desour) [reply doc/lua_api.txt]: Replaced with "whether".

File: builtin/common/tests/vector_spec.lua [hunk 10]
```diff
@@ -4,14 +4,20 @@ dofile("builtin/common/vector.lua")
...
 			local input = vector.new({ x = 3, y = 2, z = 1 })
 			local output = vector.new(input)
 			assert.same(input, output)
-			assert.are_not.equal(input, output)
+			assert.equal(input, output)
+			assert.is_false(rawequal(input, output))
+			assert.equal(input, input:new())
 		end)
 
 		it("throws on invalid input", function()
```
Comment (appgurueu) [inline builtin/common/tests/vector_spec.lua:23]: dead code?

**Comment by appgurueu** [comment]
Not a fan of the rename. If you don't want the repetition of `vector.is_vector`, cut it down to `vector.is`.

File: doc/lua_api.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (Desour) [reply doc/lua_api.txt]: I've renamed it to `is_real` now. Sounds a bit weird, but whatever.

>(BTW, I haven't forgotten about testing this -- I'm just pressed for time this week.)

(Yes, sure, no problem. : ))

**Comment by Desour** [comment]
Renamed `is_real` to `is`.

**Review by v-rob — APPROVED** [review]
Those sure are some comprehensive unit tests!  I rather doubt anything will slip through them :)

The name for `vector.is` still feels funny, but it doesn't matter a whole lot to me.

**Comment by Desour** [comment]
Squashed and rebased.
(I had to change an expected result in the unittest for `vector.round` because of #10803.)

**Comment by Desour** [comment]
With the help of #10323, you can now do `tostring(v)` to turn a vector into a string.
(Thanks for merging that PR!)

**Comment by v-rob** [comment]
Approval still stands with changes

**Comment by sorcerykid** [comment]
What about `vector.is_valid()`? The "is" by itself (at least to me, being a native English speaker) suggests a comparison, even though it's a validation. Another possibility would be "check" instead of "is_valid".

Other than that, this is shaping up to be a really nice PR!

**Comment by Desour** [comment]
`check` sounds nice, I've adopted that.

(Also squashed and rebased.)

**Comment by lhofhansl** [comment]
I like `is_valid` it's clear what it does and what it results (boolean). `check` is ambiguous (what does it check? what does it return? would it throw an exception?, etc).
Same with `is`. Nothing wrong with `'is_vector`, the meaning is clear from its name.

But I do not feel strongly, and we do not need to restart the discussion. :)

Idea is good. Lemme do a quick a review.

File: builtin/common/vector.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (appgurueu) [reply builtin/common/vector.lua:27]: Was this actually resolved?

**Comment by appgurueu** [comment]
I used the following simplified code to run some benchmarks:

```lua
local socket = require("socket")
local samples = 1e7
local assert = assert
local function bench(func)
    local time = socket.gettime()
    for _ = 1, samples do
        assert(func().z == 3)
    end
    return socket.gettime() - time
end

local metatable = {z = 3}
metatable.__index = metatable

local function new()
    return setmetatable({x = 1, y = 2}, metatable)
end

local function new_v()
    local v = setmetatable({x = 1, y = 2}, metatable)
    return v
end

local setmetatable = setmetatable
local function new_local()
    return setmetatable({x = 1, y = 2}, metatable)
end

local t1, t2, t3 = bench(new), bench(new_v), bench(new_local)

print(t1/t2, t2/t3, t1/t3)
```

Results for Lua: `0.98920939595292        0.95346128018269        0.94317285703401` => Doesn't matter much. Global call appears to be fastest however.

Results for LuaJIT: `1.0741184936436 1.9369676320273 2.0805327551495` => Global call is a bit slower than `local v`, which again is a lot slower than the localized `setmetatable`, which is roughly 2x faster than the global.

Conclusion: Include `local setmetatable = setmetatable` in your code.

**Comment by Desour** [comment]
Thanks for your idea!

I've tried to reproduce.
(I've meassured with `chronos.nanotime()` (https://luarocks.org/modules/ldrumm/chronos) because I'm not sure where "socket" comes from.)
With luajit-2.0.5 I get no significant difference. Example output:
`1.0007221295697	1.0106367633551	1.0113665740461`
With luajit-2.1.0-beta3 tough, `local v` seems to be slower than the others:
`0.56354084118264	1.7898922873673	1.0086774052493`

So, I'll use the `local setmetatable = setmetatable`.

**Comment by appgurueu** [comment]
Yay!
